Documentation/proposals/202310-automated-sharding.md

```diff
@@ -0,0 +1,164 @@
+# Shard Autoscaling
+
+* Owners:
+  * [Arthur Sens](https://github.com/ArthurSens)
+  * [Nicolas Takashi](https://github.com/nicolastakashi)
+* Related Tickets:
+  * [#4927](https://github.com/prometheus-operator/prometheus-operator/issues/4727)
+  * [#4946](https://github.com/prometheus-operator/prometheus-operator/issues/4946)
+  * [#4967](https://github.com/prometheus-operator/prometheus-operator/issues/4967)
+* Other docs:
+  * n/a
+
+This document aims to solve many small problems to unblock the usage of Horizontal Pod Autoscalers with Prometheus sharding mechanism.
+
+* [Why](#why)
+* [Pitfalls of the current solution](#pitfalls-of-the-current-solution)
+* [Goals](#goals)
+* [Non-Goals](#non-goals)
+* [Audience](#audience)
+* [How](#how)
+  * [Scale Status subresource](#scale-status-subresource)
+  * [Graceful scale-down of Prometheus Servers](#graceful-scale-down-of-prometheus-servers)
+    * [Scaling up after scaling down](#scaling-up-after-scaling-down)
+  * [Graceful scale-down of Prometheus Agents](#graceful-scale-down-of-prometheus-agents)
+* [Scale-down alternatives](#scale-down-alternatives)
+* [Action Plan](#action-plan)
+
+# Why
+
+Managing Prometheus instances with highly variable workloads can be quite challenging. When the number of scraped targets can scale up and down between a few items to thousands over a short period of time (e.g. within a day), operators are forced to opt for the safest scenario, ending up with over-provisioned instances. Additionally, the WAL replay often becomes a problem after a restart as it takes a long time to be replayed into memory. Without available replicas, this leads to unnecessary downtime.
+
+To address these issues, this document outlines a solution to enhance the Prometheus Operator by enabling the usage of Horizontal Pod Autoscalers(HPAs) to scale shards up and down. This improvement aims to simplify the process of scaling Prometheus instances, making it more accessible and efficient for operators.
+
+# Pitfalls of the current solution
+
+Prometheus can scale vertically pretty well, but instances that grow up to hundreds of GiB/TiB in WAL size start to get troublesome:
+
+* Eventual cardinality explosions have a big blast radius if a single Prometheus instance is responsible for scraping the majority of monitored applications.
+* Recovering from a crash takes several minutes due to the WAL replay.
+
+Meanwhile, another strategy would be to use Horizontal Pod Autoscalers (HPAs) with Prometheus statefulsets by increasing the number of replicas. However, arguably it makes things more complicated:
```

Comment:
(suggestion) do we need to keep this part? It should be obvious that increasing the number of replicas (not shards) doesn't solve the scalability issue. I'd rather see a brief explanation about why the actual sharding doesn't work either.

````
While the Prometheus and PrometheusAgent CRDs implement sharding for scrape configurations, there are still pending issues and missing features that prevent adoption.
````